misk/testing/src/main/kotlin/misk/testing/TemporaryFolder.kt

```diff
@@ -0,0 +1,61 @@
+package misk.testing
+
+import org.junit.jupiter.api.extension.AfterEachCallback
+import org.junit.jupiter.api.extension.BeforeEachCallback
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.api.extension.ExtensionContext
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.reflect.KMutableProperty
+import kotlin.reflect.full.createType
+import kotlin.reflect.full.isSubtypeOf
+import kotlin.reflect.full.memberProperties
+
+/** A temporary folder for use by a given test */
+class TemporaryFolder(val root: Path) {
+    /** Deletes all files and folders under the temporary folder */
+    fun delete() {
+        root.toFile().deleteRecursively()
+    }
+
+    /** @return a new folder with a randomly generated name */
+    fun newFolder(): Path = Files.createTempDirectory(root, "")
+
+    /** @return a new folder with the given name */
+    fun newFolder(name: String): Path = Files.createFile(root.resolve(name))
+
+    /** @return a new file with a randomly generated name */
+    fun newFile(): Path = Files.createTempFile(root, "", "")
+
+    /** @return a new file with the given name */
+    fun newFile(name: String): Path = Files.createFile(root.resolve(name))
+}
+
+/** Jupiter annotation that creates a temporary folder for each test, deleting after the test run */
+@Target(AnnotationTarget.CLASS)
+@ExtendWith(UsesTemporaryFolderExtension::class)
+annotation class UsesTemporaryFolder
+
+internal class UsesTemporaryFolderExtension : BeforeEachCallback, AfterEachCallback {
+    private val temporaryFolderType = TemporaryFolder::class.createType()
+
+    override fun beforeEach(context: ExtensionContext) {
+        val tempDir = Files.createTempDirectory("test-")
+        val tempFolder = TemporaryFolder(tempDir)
+        context.store("tempFolder", tempFolder)
+
+        tempDir.toFile().deleteOnExit()
+
+        @Suppress("UNCHECKED_CAST")
+        val tempFolderProperty = context.requiredTestClass.kotlin.memberProperties.firstOrNull {
+            it is KMutableProperty<*> && it.returnType.isSubtypeOf(temporaryFolderType)
+        } as? KMutableProperty<TemporaryFolder>
+
+        tempFolderProperty?.setter?.call(context.requiredTestInstance, tempFolder)
```

Comment:
hmmm . . .  this is a little weird. I like the MiskTest injecting stuff; this is more bespoke.

Reply:
And JUnit 4 rules were better!

Reply:
Agreed on Junit 4 rules.

I'll move into a `TemporaryFolderTestModule` which you can install with `MiskTest`.

I'm more and more leaning towards `MiskTest` just creating an injector, and having tests install modules which bind test callbacks for all test handling.